### DISCUSSION TITLE
Database cleanup (given day & confidence)

### DISCUSSION THREAD

**Discussion by Fish31520** [mcguirepr89/BirdNET-Pi#816] [Q&A] [answered]
Hi folks,

I would like to make a little cleanup of my detections' database.

Is there a way to delete every detection before a given day (let's say march the 24th 2023) with a confidence below 90% ?

Thanks !

**Comment by gerdesj** [comment]
How's your SQL?  Out of the box the system will maintain disc space itself.  I guess that you have changed your Minimum Confidence to 0.9 and would like to make the change across the system, including past events.

I'm a bit new around here but I imagine you'll need to run at least two SQL queries.  Go to Tools -> Database Maintenance.  Run a SELECT against detections - it's quite straight forward.  You'll need to delete the database rows and also the files, so you should run a SQL query that dumps a list of files to remove and save that, then run a DELETE FROM query on the same criteria and then use the list to delete the files.  

OK that lot isn't trivial.  You can leave the system to gradually sort itself out or you will need to learn quite a lot of stuff!

This could be a decent feature request:  A check box next to the Minimum Confidence that indicates that an increase should trigger purging of less confident data.

**Reply by Fish31520** [reply]
@gerdesj Thanks for your help !

Few months ago, @ehpersonal38 gave me a very useful command in order to delete all the entries of a given species.
It was something like
```
DELETE FROM detections WHERE Com_Name = "Northern Cardinal";
```
I guess I could modify it to something like
```
DELETE FROM detections WHERE Date <= "2023-03-24" AND WHERE Confidence < 0.9;
```
but I don't know if those arguments are correct ("<=", "AND", "0.9", etc...).

**Reply by lloydbayley** [reply]
Very-nearly! You don't need the second WHERE, so it would make it:
```
DELETE FROM detections WHERE Date <= "2023-03-24" AND Confidence < 0.9;
```
Agreeing with @gerdesj, I would do a SELECT first, which will return the data to be deleted so you can check it matches properly first. 
```
SELECT * FROM detections WHERE Date <= "2023-03-24" AND Confidence < 0.9;
```

**Reply by Fish31520** [reply]
@lloydbayley Thanks I'll try that immediately ! 😄

**Reply by Fish31520** [reply]
@gerdesj @lloydbayley @ehpersonal38 
The command ran fine but it seems like the audio files are still in ```Birdnet / BirdSongs / Extracted / By_Date```

**Reply by lloydbayley** [reply]
Ahh yes, well, that's a different kettle of fish. The SQL only removes entries from the Database. Did you run the DELETE yet or just the SELECT?

**Reply by lloydbayley** [reply]
Was just thinking about how to write a bash script to find and eradicate them but it's a mammoth task for a one-off. If you still could get the data from the SELECT statement, at least we'd have the filenames.

**Reply by gerdesj** [reply]
The filenames encode the date and confidence in their names so a careful regex should be able to get the right subset to delete if the db has already been cleared out.  

I might spend some quality time at https://regex101.com/  ...

**Reply by lloydbayley** [reply]
Yep..I was just thinking about this and about using the hyphens as delimiters but you'd have to start from the right and move left because the names of the birds can contain hyphens. Plus, the other snag is the file path. You'd have to start at the top and -s for subdirectories but as you say, be VERY careful wit the regex. It's quite a build isn't it?!

**Reply by ehpersonal38** [reply]
Just thinking, you may not need a script to delete the audio files at all, since the disk check service will eventually delete them anyway once your disk space starts getting filled up in a few weeks or so. 

Or course that assumes @Fish31520 doesn't actually care about keeping _any_ of his audio files in the long term. 

Just some food for thought 🙂

**Reply by Fish31520** [reply]
Thanks guys for your ideas and your help.

@ehpersonal38 I just moved from an appartement close to the street to a country house in the middle of nowhere, just surrounded by nature, that's why I don't really care about my "old" audio files. I just want to keep the most accurate ones and delete the others.

How does the system deal with a full hard drive ? Is it deleting audio files ? And if that's the case, on which basis (most ancient first, lowest confidence, other...) ?

Thanks for your patience !

**Reply by ehpersonal38** [reply]
If your system is set to "Purge" and not "Keep" audio files when the disk space gets > 95% used (this option is set in Advanced Settings and is the default), then BirdNET-Pi will automatically delete your old audio files, starting with the oldest ones and progressively deleting newer ones to make more space as needed. 

You can manually exclude audio clips from being deleted in this process by clicking the little lock icon at the top right of any recording. By default all your "Best Recordings" are excluded as well, but it could be interesting to automatically exclude audio files whose confidence >= 90%, for example. Of course eventually your SD card would fill up with _those_ files too and they'd eventually have to be deleted anyway, so it may just be delaying the inevitable and not that useful.

**Reply by Fish31520** [reply]
@ehpersonal38 Thanks for those explanations.

Clearly, it could be nice to purge files with lowest confidence first (and if all remaining files are above the confidence threshold then delete the oldest).

Or, I don't know, keeping the old purging system but instead of keeping only the files in the "Best recordings" page, keep the 10 files of a given species with the best confidence for example (don't know if I sound clear here ! 😅).

**Comment by gerdesj** [comment]
This is a first rough go at a regex.  It will return five capture groups - birdname, confidence, date and time and finally the file extension.

Example filename from my database: European_Robin-97-2023-03-25-birdnet-15:44:38.mp3

`(.+?)-(\d\d)-(\d\d\d\d-\d\d-\d\d)-.+?-(\d\d:\d\d:\d\d)\.(.*)`

Not properly tested, may set fire to your hair etc etc.

**Reply by lloydbayley** [reply]
Well-bloody-done on the regex and overlooking any possible dashes in the bird name. (I must admit I tried it and made your Robin a
"Farty-European_Robin-97-2023-03-25-birdnet-15:44:38.mp3"
(We know they don't exist because they would blow themselves off the branch...)

It worked fine.
Hopefully we've given @Fish31520 the makings of doing it. It's almost midnight here, so I'm not writing it! :) Hehe

**Reply by gerdesj** [reply]
It's 1400 locally (UK) so I'll crack on a bit.  Need to check the confidence can't be set to 0 or 100 or single digit.

We have put out fat balls in the past so a farty robin could easily happen and the pigeons here will eat anything 8)

**Reply by lloydbayley** [reply]
Excellent stuff. Good luck with it.

There's hilarious youtube video about pigeons that you might like. I won't post a link here but if you search for: 
"Pigeon prank call"
you'll find it.
I'm off! Hooroo

**Reply by gerdesj** [reply]
I've started poking around the source.

Any data clean up related script could start with this one: https://github.com/mcguirepr89/BirdNET-Pi/blob/main/scripts/clear_all_data.sh  The first 25 odd lines get you some handy variables.

**Reply by ehpersonal38** [reply]
Really I think modifying `disk_check.sh` might be better, around [line 22](https://github.com/mcguirepr89/BirdNET-Pi/blob/main/scripts/disk_check.sh#L22). You'd have to use your regex to match the 97 (which indicates the detection's confidence) in "European_Robin-97-2023-03-25-birdnet-15:44:38.mp3", and check if it's =< 90 before deleting the file.

**Reply by Fish31520** [reply]
@gerdesj @lloydbayley @ehpersonal38 
Thanks guys for your help !
I've had no time yet to check those ideas but I'll try to take a look in the upcoming days.
Thanks once again !

**Reply by gerdesj** [reply]
I've been poking around the source a bit when I get some time but am a long way from real familiarity.  I haven't found the routine that writes the final analyzed files yet.  

Please could you post four random filenames from your system - when it comes to date stamps it is unwise to assume anything.  It looks like dates are yyyy-mm-dd and times hh:mm:ss (24 hour clock), which is a standard (and there are loads to choose from),  but best not to assume!  Also, do you always have the string _birdnet_ in them?

Tomorrow I will be at work doing some long running tasks with multiple monitors on my PC, rather than this laptop with a dodgy power supply, a keyboard with some forgetful keys and an adequate screen.  I'll wack together a quick script to get the job done, probably safely 8)

**Reply by Fish31520** [reply]
@gerdesj

Hi,

I just saw your message (I hope, it's not too late !).

Here comes 4 random file names 😀 

- Chardonneret_élégant-80-2023-03-01-birdnet-10:19:48.mp3
- Grimpereau_des_jardins-81-2023-03-01-birdnet-18:36:37.mp3
- Mésange_charbonnière-81-2023-03-01-birdnet-09:36:33.mp3
- Pie_bavarde-80-2023-03-01-birdnet-15:34:36.mp3

They always seem to have "Birdnet" in their names...

Thanks a lot (and good luck with that laptop of yours ! 😅)

**Reply by gerdesj** [reply]
Have a look at the next comment from me.  I've been playing with "AI" 8)  I have hand edited what ChatGPT churned out - it went a bit mad at one stage but overall I'm very impressed with it.

I have not tested actual deletion because I need all my files.

**Comment by gerdesj** [comment, marked as answer]
EDIT: Whoops, forgot to deal with confidence. Back to ChatGPT ...
EDIT:  Now with added confidence handling - added by hand to avoid any off piste changes by ChatGPT

I decided to have a play with ChatGPT 8)  I ended up using Python because BASH regex is rather hard for me and ChatGPT to do and this runs a lot lot faster than my initial efforts.  This runs on an out of the box BirdNET-Pi, no additional Python modules needed.

Put the script into a file eg:

```
sudo nano /usr/local/bin/birdnet_delete_data.py 
sudo chmod +x /usr/local/bin/birdnet_delete_data.py

```
Then run it like this (date time and max confidence:
`/usr/local/birdnet_delete_data.py 2023-30-03 22:00:00 70`

By default it will only list what it wants to delete.  Add --delete to the command to actually delete the files.  If you don't provide a date, time and confidence then it will tell you what to do.  

This script does not touch your database.  That's a game for another day.

```
#!/usr/bin/env python3

import os
import re
import sys
import argparse
from datetime import datetime

# Set up argument parser
parser = argparse.ArgumentParser(description='Find all files in the BirdNET-Pi exracted directory that are earlier than a given date and time and less than a given confidence level.')
parser.add_argument('date', type=str, help='the date to match in YYYY-MM-DD format')
parser.add_argument('time', type=str, help='the time to match in HH:MM:SS format')
parser.add_argument('confidence', type=int, help='the maximum confidence value to include')
parser.add_argument('--delete', action='store_true', help='delete matched files')
args = parser.parse_args()

# Read the EXTRACTED directory from the config file
with open('/etc/birdnet/birdnet.conf') as f:
    for line in f:
        if line.startswith('EXTRACTED='):
            EXTRACTED = line.split('=')[1].strip()
            break
    else:
        sys.exit('EXTRACTED directory not found in config file')

# Append "/By_Date" to the EXTRACTED directory
EXTRACTED = os.path.join(EXTRACTED, 'By_Date')

# Assign the regex pattern, date, and time as variables
pattern = r'(.+?)-(\d\d)-(\d\d\d\d-\d\d-\d\d)-.+?-(\d\d:\d\d:\d\d)\.(.*)'
date = args.date
time = args.time
max_confidence = args.confidence

# Convert the input date and time to Unix timestamps
input_timestamp = datetime.strptime(f'{date} {time}', '%Y-%m-%d %H:%M:%S').timestamp()

# Create an empty array to hold the matched filenames
matched_files = []

# Use the find command to locate all files in the directory (including its subdirectories)
# that match the regex pattern, and loop through them
for root, dirs, files in os.walk(EXTRACTED):
    for file in files:
        # Get the filename from the full file path
        filename = os.path.join(root, file)

        # Use regex capture groups to find matches
        match = re.match(pattern, os.path.basename(filename))
        if match:
            # Extract the confidence, date and time from the filename
            file_confidence = int(match.group(2))
            file_date = match.group(3)
            file_time = match.group(4)
        
            # Convert the file date and time to Unix timestamps
            file_timestamp = datetime.strptime(f'{file_date} {file_time}', '%Y-%m-%d %H:%M:%S').timestamp()

            # Check if the file timestamp is earlier than the input timestamp and confidence level
            if file_timestamp < input_timestamp and file_confidence <= max_confidence:
                # Add the matched filename to the array
                matched_files.append(filename)

# Output and optionally delete the matched filenames
for filename in matched_files:
    print(filename)
    if args.delete:
        os.remove(filename)
```

**Reply by gerdesj** [reply]
@Fish31520 - your filenames seem to work with the regex OK \o/

**Reply by Fish31520** [reply]
@gerdesj Nice job, thanks a lot, I'm away from my RPi for few days but meanwhile I'll take a look at the code and I'll try that when I'll be back home !

Thank you and thanks to ChatGPT ! 😄

**Reply by gerdesj** [reply]
No probs.  ChatGPT is absolutely phenomenal but don't worry about it taking over the world quite yet.  It is currently a really handy calculator.  The old saying: "rubbish in rubbish out" is still very true!  Quite often it would make some quite nasty errors if not guided correctly.

It is very believable.  I found myself saying please and thank you as if I was conversing with a person.  It is way better than dealing with any search engine which are now nearly totally polluted with search engine optimisation (SEO).

The world is changing again and rather quickly.

**Reply by Fish31520** [reply]
@gerdesj 

Edit : the ^ sign is just under the "I" letter of "import re".

Hi,

Well, I finally took some time to try that 😅. Unfortunately, the command is giving me this `/usr/local/bin/birdnet_delete_data.py 2023-23-03 23:59:59 95
  File "/usr/local/bin/birdnet_delete_data.py", line 2
    import os import re import sys import argparse from datetime import datetime
        ^
SyntaxError: invalid syntax`

And also, should I use `/usr/local/birdnet_delete_data.py YYYY-DD-MM 23:59:59 95` or rather a YYYY-MM-DD pattern ?

Thanks !

**Reply by gerdesj** [reply]
Call it with YYYY-MM-DD HH:MM::SS which is the format the files use.  I think that is an ISO/OSI format of some sort.

I've just checked and re is part of the standard Python library so should be installed.  The error is on line two which is a blank line in my script, import re is on line 4.  Please double check you've pasted it or transferred it OK.  If you used a Windows box to create the file and then transferred it over it will have some odd line endings!

Let us know how you get on.

**Reply by jflyons** [reply]
Appreciate the discussion and the mention of using ChatGPT.  I have been using, though for much less ambitious coding/spreadsheet ideas, so far anyway.

**Reply by Fish31520** [reply]
@gerdesj 

"If you used a Windows box to create the file and then transferred it over it will have some odd line endings!"

Don't know if that's your point but I'm using the Web Terminal...

And you're right, when I copy/paste the script you wrote, Nano refuse to paste it in one piece, I've got to make the job in 3 steps. It's also changing the spacing between words/lines...

Edit : I used ssh to access Birdnet & Nano and it worked pretty well. The script was copied correctly and the command worked smoothly (erased 10Gb from my oldest files, keeping only the ones with the best confidence).

Thanks a lot for your help !

**Reply by Fish31520** [reply]
@jflyons
Well, ChatGPT is no magic but it seems to be pretty usefull when you ask it precise questions. It's a nice tool.

**Reply by gerdesj** [reply]
@Fish31520 - my pleasure.
